rre-core/src/main/java/io/sease/rre/core/domain/metrics/impl/NDCGAtK.java

```diff
@@ -0,0 +1,129 @@
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.StreamSupport;
+
+import static io.sease.rre.Func.gainOrRatingNode;
+import static java.util.stream.Collectors.groupingBy;
+
+/**
+ * NDCG@k metric.
+ *
+ * @author agazzarini
+ * @since 1.0
+ */
+public class NDCGAtK extends Metric {
+    private final static BigDecimal TWO = new BigDecimal(2);
+    final int k;
+
+    /**
+     * Builds a new NDCGAtK metric.
+     *
+     * @param k the top k reference elements used for building the measure.
+     */
+    public NDCGAtK(@JsonProperty("k") final int k) {
+        super("NDCG@" + k);
+        this.k = k;
+    }
+
+    @Override
+    public int getRequiredResults() {
+        return k;
+    }
+
+    @Override
+    public ValueFactory createValueFactory(final String version) {
+        return new ValueFactory(this, version) {
+            private BigDecimal dcg = BigDecimal.ZERO;
+
+            @Override
+            public void collect(final Map<String, Object> hit, final int rank, final String version) {
+                if (rank > k) return;
+                judgment(id(hit))
+                        .ifPresent(judgment -> {
+                            final BigDecimal value = gainOrRatingNode(judgment).map(JsonNode::decimalValue).orElse(TWO);
+                            BigDecimal numerator = TWO.pow(value.intValue()).subtract(BigDecimal.ONE);
+                            if (rank == 1) {
+                                dcg = numerator;
+                            } else {
+                                double den = Math.log(rank + 1) / Math.log(2);
+                                dcg = dcg.add(numerator.divide(new BigDecimal(den), 2, RoundingMode.FLOOR));
+                            }
+                        });
+            }
+
+            @Override
+            public BigDecimal value() {
+                if (totalHits == 0) {
+                    return relevantDocuments.size() == 0 ? BigDecimal.ONE : BigDecimal.ZERO;
+                }
+
+                final BigDecimal idealDcg = idealDcg(relevantDocuments);
+                if (dcg.equals(BigDecimal.ZERO) && idealDcg.equals(BigDecimal.ZERO)) {
+                    return BigDecimal.ZERO;
+                }
+
+                return dcg.divide(idealDcg, 2, RoundingMode.FLOOR);
+            }
+        };
+    }
+
+    private BigDecimal idealDcg(final JsonNode relevantDocuments) {
+        final int windowSize = Math.min(relevantDocuments.size(), k);
+        final int[] gains = new int[windowSize];
+
+        final Map<Integer, List<JsonNode>> groups =
+                StreamSupport.stream(relevantDocuments.spliterator(), false)
+                                .collect(groupingBy(doc -> gainOrRatingNode(doc).map(JsonNode::intValue).orElse(2)));
```

Comment:
what is 2 here ?
Should it be replaced with a constant, that gives an hint to what it means?
Is it the same 2 of the missing rating?
"defaultRating, with a comment that says, this is the default rating used if the search result has no rating?"

Furthermore is not very readable, what does split iterator do ?

Reply:
I've added a constant DEFAULT_RATING